x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java

```diff
@@ -720,7 +720,9 @@ public JoinInfo visitJoinCondition(EsqlBaseParser.JoinConditionContext ctx) {
             throw new ParsingException(source(ctx), "JOIN ON clause cannot be empty");
         }
 
-        // inspect the first expression to determine the type of join (field-based or expression-based)
+        // Inspect the first expression to determine the type of join (field-based or expression-based)
+        // We treat literals as field-based as it is more likely the user was trying to write a field name
+        // and so the field based error message is more helpful
```

Comment:
Ah! Thanks, I would not have thought that the error message is the reason for checking for literals here :)